### DISCUSSION TITLE
Release Candidate #8

### DISCUSSION THREAD

**Discussion by DeeEmm** [DIY-Flow-Bench/DIY-Flow-Bench#189] [Code Development]
RC8 is a major step towards the final release. There a number of bugs that have been addressed as well as a large number of features added. Many of the changes would not have been possible without the involvement of the DIYFB community who have been been instrumental with assisting with both bug fixes and functionality.

Major features include
- Support for differential pressure style benches added
- GUI Firmware updates now possible
- Rewrite of calibration code

This release is our eighth release candidate and is fully functional code that is released with the intention of peer testing. Once the peer testing process has finished and any bugs addressed, the final release candidate will become the official V2.0 release.

The main focus of units tests should now be accuracy and repeatability.

Alongside this release we have also validated the new V2.3 shield PCB which will be available to purchase shortly.

Releases will now be packaged up with Binaries for the Wemos D1 R32 ESP32 which is the officially recommended board for the project. These binaries should work for ALL ESP32's that compile under the standard unmodified platformio.ini [building using release > esp32dev]

To be able to update firmware in the GUI, you will first need to install this release using your regular IDE. You will also need to update your index.html file as well.

For discussions relating to this release, including bug reports please visit the official thread in the forums

```
Build #         - Description of Change

24110201        - SD Card methods validated 
                - Missing line terminator #187 
                - Changed unused pins to '-1' in pins.h
24110101        - Missing line terminator #187 
                - Added user definable flow differential target
                - Finalised flow differential code
24103002        - Added additional 'Flow Differential' tile
24103101        - Updated leak cal fucntionality as per discssion #178
                - Refactored webserver class to reduce spaghetti
                - General code tidying
24103002        - Reformatted BOM CSV for Git rendering
24103001        - Added manual save function for calibration data
24103001        - Added firmware update via GUI
24102902        - Updated README
24102901        - Fixed bug where filter value was not retained
24102803        - Added data filter for Rolling Median and base code for Cyclic Average and Mode
24102802        - Added MAF Flow tile to dashboard
24102801        - Updated flow calibration routine
                - Changed partition regime to allow for OTA
                - General code tidy
                - Updated language constants
                - Fixed issue with index.html upload
24102701        - Moved calibration buttons to main tab
24102601        - #178 - Leak Test Calibration data 
                - #179 - Flow test Calibration
24092301        - Added basic support for orifice / pitot / venturi style benches
```

### What's Changed?
* Build 24103001        - Added firmware update via GUI by @DeeEmm in https://github.com/DeeEmm/DIY-Flow-Bench/pull/185
* Build 24110201 by @DeeEmm in https://github.com/DeeEmm/DIY-Flow-Bench/pull/188

**Full Changelog**: https://github.com/DeeEmm/DIY-Flow-Bench/compare/V2.0-RC.7...V2.0-RC.8

<hr /><em>This discussion was created from the release <a href='https://github.com/DeeEmm/DIY-Flow-Bench/releases/tag/V2.0-RC.8'>Release Candidate #8</a>.</em>

----------------------------------------

### UPDATE: RC.8 issues to be fixed for RC.9 Release...

https://github.com/DeeEmm/DIY-Flow-Bench/issues?q=is%3Aopen+is%3Aissue+milestone%3ADIYFB-V.2.0-RC.9

**Comment by black-top** [comment]
The "GUI Firmware updates now possible" is great feature, but I think we need the ability to select all sensors at GUI configuration because firmware.bin  files will be compiled with all default sensors ?

**Reply by stefan63** [reply]
I aggree and are thinking maybe an alternative could be to omit the configuration.h file from upgrade files.
And make it possible to upload this file separate in UI if config need to be changed?

**Reply by DeeEmm** [reply]
I have considered trying to make the configuration file uploadable via the GUI. The only issue there is that at present the configuration file drives many pre-compile directives, which it obviously cannot do post-compile.

It would be possible to get rid of the pre-compile directives and just include everything within the compiled code. This would increase file size, but I don't think this would be an issue. It would however be a major re-write.

The configuration file would need to be changed into JSON format. It would then be handled the same as existing calibration and config files. This would also allow for pre-canned configuration files that cater for different hardware setups. 

- load the base firmware.
- load the index.html file
- load the config file for your setup

You could theoretically be up and running without ever changing a setting.

Furthermore if we include a browser based editor, then you can even edit the config file within the browser. This saves having to hard code configuration editing into the existing GUI. But I do think that there is value in moving some items into the GUI config settings, such as the trimpot values,

**Reply by DeeEmm** [reply]
Added #195

**Comment by DeeEmm** [comment]
We've made a few updates to RC8 already, which means that there will inevitably be an RC9 release.

The project has always suffered from a bit of scope creep, which is absolutely fine of course, sometimes it has given me things to do whilst I wait for feedback. But I feel that I would like to try and put a line in the sand and lock down the features to concentrate on getting the actual V2 release out.

So the project roadmap is already defined. V2 should really just be about validation. This of course includes all bug fixes found along the way.

https://github.com/users/DeeEmm/projects/5

I am happy to roll up additional functionality within the V2 release, if we think that it adds value to the **basic** project. When I say '_basic_' I mean the ability to accurately measure and display flow. At present we have not validated the accuracy of the project and this is where I feel we now need to concentrate. We can still capture all ideas along the way and then review them afterwards. 

Happy to open up the discussion on what everyone else would like to see.

I've pinned issue #35 [Validation Tests] as I think that this is where we need to concentrate. To come up with a series of validation tests and then compare results across several different benches

**Comment by DeeEmm** [comment]
So I've ended up down the rabbit hole a little taking a look at loading configuration after compile time. Initially I wanted to see if it was feasible (it is), and so did a proof of concept by re-writing the pins file and the way that the I/O is handled.

We've moved from a precompile #define...

```
    #define FLOW_VALVE_STEP_PIN         32
    #define FLOW_VALVE_DIR_PIN          33
```

to storing the data in a JSON file

```
{
    "BOARD_TYPE" : "WEMOS_D1_R32",
    "FLOW_VALVE_STEP_PIN" : 32,
    "FLOW_VALVE_DIR_PIN" : 33,
   ...
}
```

And then loading the data into memory at run time

```
    pinData["FLOW_VALVE_STEP_PIN"].as<int>() != -1 ? pinMode(pinData["FLOW_VALVE_STEP_PIN"].as<int>(), OUTPUT) :  void() ;
    pinData["FLOW_VALVE_DIR_PIN"].as<int>() != -1 ? pinMode(pinData["FLOW_VALVE_DIR_PIN"].as<int>(), OUTPUT) :  void() ;
```

This retains the methodology from @KirikauKiwi where unused pins are not assigned. We retain the '-1' nomenclature for unused pins

The difference between the files is subtle and the resultant file is as easy to read, however JSON does not allow comments as it is a data-only format and without the comments there is a bit information overload.

We can add comments to a version on the WIKI, which will provide a point of reference, and move some of the more commonly accessed variables out into the configuration tab (i.e. trimpots).

At this point in time I've got pins.json working and have converted the configuration.h file into JSON format. There is still an amount of work to do, to go through the code and replace the existing #ifdefs, (only 39 instances). I'm kinda thinking that I've started so might as well finish.

Totally sidetracked lol

pins.json
```
{
    "BOARD_TYPE" : "WEMOS_D1_R32",
    "VAC_SPEED_PIN" : 25, 
    "VAC_BLEED_VALVE_PIN" : 26,
    "VAC_BANK_1_PIN" : 13,
    "VAC_BANK_2_PIN" :12, 
    "VAC_BANK_3_PIN" :14,
    "AVO_STEP_PIN" :  15,
    "AVO_DIR_PIN" : 27,
    "FLOW_VALVE_STEP_PIN" : 32,
    "FLOW_VALVE_DIR_PIN" : 33,
    "VCC_3V3_PIN" : -1,
    "VCC_5V_PIN" : 35,
    "SPEED_SENSOR_PIN" : 2,
    "SWIRL_ENCODER_PIN_A" : 2,
    "SWIRL_ENCODER_PIN_B" : 4,
    "ORIFICE_BCD_BIT1_PIN" : 34,
    "ORIFICE_BCD_BIT2_PIN" : 36,
    "ORIFICE_BCD_BIT3_PIN" : 39,
    "MAF_PIN" : -1,
    "REF_PRESSURE_PIN" : -1,
    "DIFF_PRESSURE_PIN" : -1,
    "PITOT_PIN" : -1,
    "TEMPERATURE_PIN" : -1,
    "REF_BARO_PIN" : -1,
    "HUMIDITY_PIN" : -1,
    "SERIAL0_TX_PIN" :1,
    "SERIAL0_RX_PIN" :3,
    "SERIAL2_TX_PIN" :16,
    "SERIAL2_RX_PIN" :17,
    "SDA_PIN" : 21,
    "SCL_PIN" : 22,
    "SD_CS_PIN" : 5,
    "SD_MOSI_PIN" : 23,
    "SD_MISO_PIN" : 19,             
    "SD_SCK_PIN" : 18,
    "WEMOS_SPARE_PIN_1": 4
}
```

configuration.json
```
{
"FORMAT_FILESYSTEM_IF_FAILED" : true,
"SD_CARD_IS_ENABLED" : false,
"SERIAL0_ENABLED" : true,
"SERIAL2_ENABLED" : false,
"SERIAL0_BAUD" : 115200,
"SERIAL2_BAUD" : 9600,
"CUSTOM_MAC_ADDRESS" : false,
"MAC_ADDRESS" : "{0x32, 0xAE, 0xA4, 0x07, 0x0D, 0x00}",
"STATIC_IP_ADDRESS" : false,
"STATIC_IP" :  "{192,168,1,222}",
"SUBNET" : "{192,168,1,1}",
"GATEWAY" : "{255,255,0,0}",
"WEBSERVER_ENABLED" : true,
"BOOT_MESSAGE" : "May the flow be with you...",
"PAGE_TITLE" : "DIY Flow Bench",
"LANGUAGE_FILE" : "language/EN_language.h",
"SHOW_ALARMS" : true,
"MIN_REFRESH_RATE" :  250,
"API_IS_ENABLED" : true,
"API_BLOB_LENGTH" : 1024,
"API_RESPONSE_LENGTH" : 64,
"API_STATUS_LENGTH" : 128,
"API_JSON_LENGTH" : 1020,
"API_SCAN_DELAY_MS" : 250,
"PRINT_BUFFER_LENGTH" : 128,
"API_CHECKSUM_IS_ENABLED" : false,
"FILESYSTEM" :  "SPIFFS",
"SENSOR_TASK_MEM_STACK" : 2200,
"ENVIRO_TASK_MEM_STACK" : 1800,
"DATA_JSON_SIZE" : 1500,
"CONFIG_JSON_SIZE" : 1536, 
"CAL_DATA_JSON_SIZE" : 128,
"LIFT_DATA_JSON_SIZE" : 384,
"VTASK_DELAY_ADC" : 500,
"VTASK_DELAY_BME" : 500,
"VTASK_DELAY_SSE" : 500,
"STATUS_UPDATE_RATE" : 500,
"MIN_TEST_PRESSURE_PERCENTAGE" : 80,
"PIPE_RADIUS_IN_FEET" : 0.328084,
"VCC_3V3_TRIMPOT" : 0.0,
"VCC_5V_TRIMPOT" : 0.0,
"USE_FIXED_3_3V_VALUE" : false,
"USE_FIXED_5V_VALUE" : false,
"BME_IS_ENABLED" : true,
"BME280_I2C_ADDR" : "0x76",
"BME_SCAN_DELAY_MS" :  1000,
"ADC_IS_ENABLED" : true,
"ADC_I2C_ADDR" : "0x48", 
"ADC_SCAN_DELAY_MS" : 250,
"ADC_MAX_RETRIES" : 10,
"ADC_RANGE" : 32767,
"ADC_GAIN" : 6.144,
"ADC_TYPE_ADS1115": true,
"MAF_IS_ENABLED" : true,
"MAF_DATA_FILE" : "mafData/BOSCH_0280218067.cpp",
"MAF_SRC_IS_PIN" : false,
"MAF_SRC_IS_ADC" : true,
"MAF_MV_TRIMPOT" : 0.0,
"MAF_ADC_CHANNEL" : 0,
"PREF_IS_ENABLED" : true,
"FIXED_REF_PRESS_VALUE" : 1,
"PREF_SRC_PIN" : false,
"PREF_SRC_ADC" : true,
"PREF_SENSOR_TYPE_LINEAR_ANALOG": false,
"PREF_SENSOR_TYPE_MPXV7007" : true,
"PREF_SENSOR_TYPE_MPXV7025" : false,
"PREF_SENSOR_TYPE_XGZP6899A007KPDPN" : false,
"PREF_SENSOR_TYPE_XGZP6899A010KPDPN" : false,
"PREF_SENSOR_TYPE_M5STACK_TubePressure" : false,
"PREF_MV_TRIMPOT" :  0.0,
"PREF_ANALOG_SCALE" :  1.0,
"PREF_ADC_CHANNEL" : 1,
"PDIFF_IS_ENABLED" : true,
"FIXED_DIFF_PRESS_VALUE" : 1,
"PDIFF_SRC_IS_PIN" : false,
"PDIFF_SRC_IS_ADC" : true,
"PDIFF_SENSOR_TYPE_LINEAR_ANALOG" : false, 
"PDIFF_SENSOR_TYPE_MPXV7007" : true, 
"PDIFF_SENSOR_TYPE_MPXV7025" : false,
"PDIFF_SENSOR_TYPE_XGZP6899A007KPDPN" : false,
"PDIFF_SENSOR_TYPE_XGZP6899A010KPDPN" : false,
"PDIFF_SENSOR_TYPE_M5STACK_TubePressure" : false,
"PDIFF_MV_TRIMPOT" : 0.0,
"PDIFF_ANALOG_SCALE" : 1.0,
"PDIFF_ADC_CHANNEL" : 2,
"PITOT_IS_ENABLED" : true,
"PITOT_SRC_IS_PIN" : false,
"PITOT_SRC_IS_ADC" : true,
"PITOT_SENSOR_NOT_USED" : false,
"PITOT_SENSOR_TYPE_LINEAR_ANALOG" : false,
"PITOT_SENSOR_TYPE_MPXV7007" : true,
"PITOT_SENSOR_TYPE_MPXV7025" : false,
"PITOT_SENSOR_TYPE_XGZP6899A007KPDPN" : false,
"PITOT_SENSOR_TYPE_XGZP6899A010KPDPN" : false,
"PITOT_SENSOR_TYPE_M5STACK_TubePressure" : false,
"PITOT_MV_TRIMPOT" : 0.0,
"PITOT_ANALOG_SCALE" : 1.0,
"PITOT_ADC_CHANNEL" : 3,
"BARO_IS_ENABLED" : true,
"BARO_SENSOR_TYPE_FIXED_VALUE" : false,
"BARO_SENSOR_TYPE_LINEAR_ANALOG" : false,
"BARO_SENSOR_TYPE_BME280" : false,
"BARO_SENSOR_TYPE_MPX4115" : false,
"FIXED_BARO_VALUE": 101.3529,
"BARO_ANALOG_SCALE" : 1.0,
"startupBaroScalingFactor" : 1,
"startupBaroScalingOffset" : 100,
"BARO_MV_TRIMPOT" : 0.0,
"BARO_FINE_ADJUST" : 0.0,
"SEALEVELPRESSURE_HPA" : 1016.90,
"TEMP_IS_ENABLED" : true,
"FIXED_TEMP_VALUE" : 21,
"TEMP_ANALOG_SCALE" : 1.0,
"TEMP_SENSOR_NOT_USED" : false,
"TEMP_SENSOR_TYPE_FIXED_VALUE" : false,
"TEMP_SENSOR_TYPE_LINEAR_ANALOG" : false,
"TEMP_SENSOR_TYPE_BME280" : true,
"TEMP_SENSOR_TYPE_SIMPLE_TEMP_DHT11" : false,
"TEMP_MV_TRIMPOT" : 0.0,
"TEMP_FINE_ADJUST" : 0.0,
"RELH_IS_ENABLED" :true,
"FIXED_RELH_VALUE" : 36,
"RELH_ANALOG_SCALE" : 1.0,
"RELH_SENSOR_TYPE_FIXED_VALUE" : false,
"RELH_SENSOR_TYPE_LINEAR_ANALOG" : false,
"RELH_SENSOR_TYPE_BME280" : true,
"RELH_SENSOR_TYPE_SIMPLE_RELH_DHT11" : false,
"RELH_MV_TRIMPOT" : 0.0,
"RELH_FINE_ADJUST" : 0.0,
"SWIRL_IS_ENABLED" :false
}
```

**Reply by DeeEmm** [reply]
I think that some of these can be changed.

For example we have several definitions where we comment out all but the definition that we want. This doesn't work that well in JSON as we have a bunch of replication...

```
"PITOT_SENSOR_TYPE_LINEAR_ANALOG" : false,
"PITOT_SENSOR_TYPE_MPXV7007" : true,
"PITOT_SENSOR_TYPE_MPXV7025" : false,
"PITOT_SENSOR_TYPE_XGZP6899A007KPDPN" : false,
"PITOT_SENSOR_TYPE_XGZP6899A010KPDPN" : false,
"PITOT_SENSOR_TYPE_M5STACK_TubePressure" : false,
```

These could be replaced with 

```
"PITOT_SENSOR_TYPE" : "MPXV7007",
```

But this does require that the end user locates the correct value from the WIKI and more coding is required to decode the value

This only needs to be done once. Then on subsequent updates, only the binary firmware file will need to be uploaded. All exiting pin definitions and configuration settings will be retained.

**Reply by DeeEmm** [reply]
Current version is in the WIP branch. 

You will need to create a pins.json file from the example WEMOS_D1_R32.json in the pins folder

**Reply by black-top** [reply]
It looks like a deep rabbit hole. I  don't know how it is currently implemented in the settings tab of the drop-down menu for bench types and, for example, temperature units. Maybe it would be easier to do the same? Just select sensors and pins from the drop-down menu ?

**Comment by black-top** [comment]
BTW, it looks like after some updates, the lift points at data capture are no longer possible to change?

**Comment by stefan63** [comment]
Good work the last weeks!
GUI updates and GUI configuration and settings id the way to go to gain more users and interest in this project!
Then add an easy way to load the initiol firmware to the ESP (Take a look at FluidNC and grblHAL web uploaders) and this is a killer application. 
Ref: 
**grblHAL web builder**  "http://svn.io-engineering.com:8080/"
**FluidNC Web Installer** "https://installer.fluidnc.com/"

**Comment by DeeEmm** [comment]
It's a fair bit of work to get to a dropdown for each setting. First we need to get all of the settings out into dedicated files and then we can look to create GUI elements to manage each setting. Ultimately down the track something like this would be the goal but for the time being just the ability to change settings post compile time will allow the OTA firmware to function correctly.

It's quite interesting changing the settings from pre-compile to post-compile. It presents a number of technical challenges. 

We need to change the following files to be loaded after the ECU has booted

- pins.h  - pins definitions + I/O assignments
- configuration.h - configuration settings > begin() & initialise() vars setup
- maf.cpp - MAF data creation
- language.h language file 

Changing to a JSON format and reading the variables in from a file is not too hard as we already use the arduinoJSON library which makes relatively easy work of this, most of the settings are easy enough to translate apart from the large data array that is used for the MAF file, which I have yet to do. There is some creative coding needed to ensure that we can typecast correctly from char* arrays to whatever variable type we require to get the settings from JSON. This typecasting was previously taken care of automagically by the #define precompile directive and was transparent. So there is an amount of work to do there. Also char* / string management in C++ is woeful. Give me PHP any day lol.

There are also some additional challenges depending where we need to replace compile time directives within the code. Previously we #include the files at the top of each class or file as required. This meant that the values from those files were available right away to use. However when we have some #ifdef statements in the header section of the class definitions, we cannot use regular conditional logic to replace them, as we cannot use if-then-else logic in the header section of the file. So some creative solutions are required.

We also need to bypass pretty much all of the code until the I/O and configuration files are present. So we need to be able to block out large portions of the code so that it skips directly to the webserver to allow the relevant files to be uploaded. The big challenge there is dealing with unassigned I/O and variables (because we've essentially removed most vars from the code) and so we have to bypass the code that relies on them whilst still allowing it to function afterwards. The existing code flow is not ideal for this as we did all of the I/O and configuration stuff first, whereas we are now trying to do it last.

Today has been working through many errors line by line to get the code to compile without having the configuration and I/O assigned.

Very challenging lol.

**Reply by KirikauKiwi** [reply]
That must be a big task indeed as it is a major restructure. I assume this is a big WIP currently and is it something more in keeping with V3 of the DIYFB project?
Appreciate the updates.

**Reply by DeeEmm** [reply]
Yeah it's something I wished I'd not started lol. One small change that leads on to many. I'll see if I can get it together, otherwise I might just park it and revisit it another time. Might be a V5 thing lol

Moving the vars from the #includes to JSON files is pretty much done, it's handling not having the var data in the main code that's proving to be the big challenge. It's throwing up some curveballs. To the point that I've had to dig out the debugger to see what is going on. lol.

**Reply by DeeEmm** [reply]
I've parked the post compile time stuff for the time being. But it has been an interesting exercise and has given some valuable insight into what needs to happen when I get back to it.

The problem with updates like this is that they affect all files and require many concurrent changes, so there is little opportunity to stage the changes, and implement them bit-by-bit, it is essentially all or nothing. 

There is however opportunity to slowly change the code philosophy to make the eventual changes a little easier, and that is probably the most valuable take-away from this.

**Comment by DeeEmm** [comment]
Also.... Cannot find the discussion about the valve lift capture not working. I shot this screencast to show workflow.

https://github.com/user-attachments/assets/78303aff-b440-4202-8d57-eb95daf6d205

Functionality is only very basic at the moment. Will look to expand that in V3

**Reply by black-top** [reply]
Yes, but the lift points themselves, I think, were possible to change before or not?

**Reply by DeeEmm** [reply]
Yes. Within the configuration settings you have an option for 'Valve lift interval'. This sets the 'pitch' of the valve lift.

<img width="834" alt="Screenshot 2024-11-08 at 11 44 55 pm" src="https://github.com/user-attachments/assets/6dee11c4-62b1-4a26-882b-67610ef97ece">

There are 12 valve lift data-points in total. 
If you set the interval to 1.50 you will end up with the lift range shown above. i.e. 12 x 1.5 
e.g. 1.5 | 3 | 4.5 |  6 | 7.5 ... 16.5 | 18 mm
Or you can change this to whatever suits your setup.
For example. On an imperial system you might find an interval of  0.05" better, as this will give you  a total range of approx 600 thou

**Reply by KirikauKiwi** [reply]
Is it possible on the GUI configuration page to add a "unit" label beside each of the drop downs.
Example for the Valve Lift interval it would be show "mm" or "inches/thou"?

**Reply by black-top** [reply]
Sorry, all good. Don't know how I didn't notice that setting.

**Reply by DeeEmm** [reply]
> Is it possible on the GUI configuration page to add a "unit" label beside each of the drop downs. Example for the Valve Lift interval it would be show "mm" or "inches/thou"?

At the moment the unit is completely agnostic, it is neither mm or inches but both. This is intentional as it allows it to cater for all denominations without having to add a bunch of extra code to control how the units is displayed in the GUI. 

But we could add '(mm / ")' or '(mm / inch)' or something similar to the config item to make it less ambiguous.

I do want to revise some of the other descriptions to make them less ambiguous. For example for 'Conversion Depression' I want to change it to something like  'Adjusted Flow Depression'. 

If there are any other suggestions, let me know

**Reply by black-top** [reply]
Together with valve lift interval can be added maximum flow value for the graph scaling.

**Reply by DeeEmm** [reply]
Currently the scaling is automatically set based on the max flow value of the MAF.  It will set max scale at either 1000cfm, 500cfm or 250cfm and automatically scale the other values.

```
  // very rough cfm calculation from max mafdata (approx half of KG/h rate)
  if (status.mafUnits == MG_S) {
    maxcfm = (0.0036 * status.mafDataValMax) / 2; 
  } else {
    maxcfm = status.mafDataValMax / 2;
  }

  // Determine data graph flow axis scale
  // NOTE: currently 1000cfm is the largest flow that the graph will display. 
  // We could change scaling to be realtive to SVG height (surface currently fixed at 500)
  if (maxcfm < 500) {
    maxval = 250;
    scaleFactor = 2;
  } else if (maxcfm > 250 && maxcfm < 500) {
    maxval = 500;
    scaleFactor = 1;
  } else if (maxcfm > 500){
    maxval = 1000; 
    scaleFactor = 0.5;
  }
```

**Reply by black-top** [reply]
Okay, that makes sense. It just looked a bit out of proportion with 1000CFM, not many heads flow that much. :D

**Reply by DeeEmm** [reply]
I guess it would be good to override it if for example you have a large MAF but only typically flow heads up to 300cfm.

I will have a ponder about how to override this in the config

**Comment by DeeEmm** [comment]
I've added units to each of the items...

<img width="760" alt="Screenshot 2024-11-09 at 11 31 13 am" src="https://github.com/user-attachments/assets/a6b997ca-0802-40a3-a453-6413b0bc206c">

**Comment by black-top** [comment]
I'm not sure if we need a separate topic for close loop control. But just a few thoughts here. With a DBW setup, it might be difficult to control small variations, probably very small diameter DBW will be necessary. I'm thinking of testing a relatively large idle valve, it might be more precise, and most of them are controlled by PWM, but for PWM output, it will be good to have frequency adjustment because a lot of valves are frequency sensitive. DBW, probably best for setups without VFD, that pump runs 100% all the time.

**Reply by DeeEmm** [reply]
I've got a DBW throttle body here that I have done some preliminary testing with. But not hooked up in closed loop at this stage. If it does prove to be too large then it might be better to use a throttle idle valve. I have a bunch of those here too, both stepper and PWM control.

Definitely worth separating those discussions into a seperate thread.

**Comment by DeeEmm** [comment]
OK, so I'm trying to smash through any outstanding bug reports. I also want to make sure that any feature that is currently present in the GUI is functional, so this includes pitot and swirl. I feel that this is a good place to feature lock V2. If it's already present it should work, and if its not already present, it gets captured as a feature request or an enhancement, and will be reviewed as part of the V3 development.

I also have to undertake a functional test of the stepper drivers so that I can release the PCB. I've had a lot of enquiries so I think that these two things together - V2 finalisation and PCB release will be good for the project. I need to get to a position where I can ship the PCBs

So I've created a project for RC9 so that we can manage what needs to be done.

https://github.com/users/DeeEmm/projects/9/views/1

I will populate it shortly

**Comment by DeeEmm** [comment]
Right. I've reviewed all open issues and added all of those to the roadmap for RC9 that meet the 'if its there already it should work' criteria.

I've also added a few additional things that I need to get done. Please review the list and chime in with anything that I've missed or is not currently captured but should be included.

**Reply by black-top** [reply]
The basic version of PWM close loop pressure control.

**Reply by KirikauKiwi** [reply]
BME68X incorporated in to the code please.

**Reply by KirikauKiwi** [reply]
Piggy backing on the PWM of @black-top & @stefan63 but for motor control, I have ordered the [Kemo M028N & M015N](https://www.kemo-electronic.de/en/Light-Sound/Effects/Modules/M028N-Power-control-110-240-V-AC-4000-VA.php) to try out on the vac motors, it should be here in a day or so.

**Reply by DeeEmm** [reply]
Ahh yes, forgot about the BHE68X. Have one here to test with.

**Reply by DeeEmm** [reply]
#233 created for BME68X

**Reply by DeeEmm** [reply]
> The basic version of PWM close loop pressure control.

I think this is really a V3 item. There is currently no closed loop control in the code so it is a new feature.

It will still (potentially) get addressed within the same timeframe as I suspect that this will be one of the higher priority V3 tasks.

Plan is once V2 is done we will prioritise V3 tasks and look to have a regular release cycle for feature updates.

**Comment by DeeEmm** [comment]
So we've made some good progress towards RC9. We are now down to 9 open issues...

<img width="1256" alt="Screenshot 2024-11-23 at 12 14 28 am" src="https://github.com/user-attachments/assets/df8785d3-1548-4640-a669-cc2d6fd6eccb">

I think that possibly one or two of those might also be ready to close out too.

Hopefully we could be looking at RC9 in the week or two.

Great work guys. Thanks for all of your support. 👍

**Comment by DeeEmm** [comment]
Merged the DEv branch into master to allow me to sort out the auto branch creation workflow. What was quite astounding was the changelog diffs from the last master update...

```
24112402        - #213 Pitot / differential zeroing
                - #254 Report sensor voltage to API
                - Cookie to remember tile status
24112401        - #158 Correction to output
24112301        - #158 Add capture datatype selection
24112203        - #212 Changed Modal Headers
24112202        - #212 Change info modal to two column layout
24112201        - #244 simplification of resolution / accuracy
24112102        - Updated pins files
24112101        - #213 Working pitot velocity
                - Zero reset for pitot and PDiff sensors
                - Added pitot velocity calcs
                - Typecast pins to pin struct from pins.json
                - Change comparitor for testing pins in initialiseIO 
                - Added additional checks to initialiseIO
24112001        - Added Language Template vars to GUI
24111901        - Added Create-Issue-Branch workflow
                - General project refactor
24111801        - Added BME680 Configuration
                - Forked BME680 library (https://github.com/DeeEmm/BME680)
                - Moved github documentation into .github folder
24111702        - #213 Pitot changes
                - GUI Updated
                - Added help prompt to serial monitor boot sequence
                - Stepper Unit tests
                - Updated PINS_WEMOS.json
                - Updated pins loading and assignment 
24111701        - #237 Update pin assignmnt
24111601        - Code Re-factor
                - Added check to pins initialisation
24111501        - Added pins.json files
                - added 'QuickStart.md' to Docs
                - Updated contributing.md
                - Created getRemote() function to read remote file
                - Allow non-specific file name for pins and maf uploads
24111401        - GUI Updated
                - Added GUI version to info pane
                - Moved system settings into system.h 
                - Split webserver methods out into comms / data / server functions
                - Added pre-compile build vars for version in main code
                - Tidied platformio.ini
                - Added auto version for main code
                - created user-actions.py for compile time tasks (requires python installed)
24111201        - index.html versioning
                - Move data related methods into dedicated datahandler class
                - Move pins.h into JSON file
                - Added index.html versioning
```

And that's not the only changes since RC8 was released, there were already changes present on master...

```
24111107        - #222 Adjusted standard flow
24111106        - #220 Datagraph max value
24111105        - #212 Added scroll to modals
24111104        - #217 add GPT version
24111103        - #217 Scaled MAF housing returns O.O.R. value
24111102        - #212 Modal Window cropped
24111101        - #208 Added SCFM
                - #212 Display issues
                - Redirect to Main after config change
24110901        - Changed the pressure ratio formula to allow +/-ve values to be calculated
24110802        - #207 Decimal places on GUI
24110801        - Code tidy up so the pressure sensors are correctly reported in the web GUI info page
24110704        - Updated calls to // #include " version.h"
24110703        - Added call to // version.h in webserver.h
24110702        - Changed min pressure threshold to allow -ve values
                - Moved version vars into dedicated file
24110701        - M5Stack Tube pressure sensors initialisation
                - Correction to MAF rescaling function
24110403        - Added HTTP API Call for JSON Data
24110402        - #180 Added ability to truncate / round flow values in GUI
24110401        - #179 - Incorrect convertFlowDepression method
24110302        - Fixed flow thresholds to allow negative flow values in GUI
24110301        - Added M5Stack Core2 Support
                - Added M5Stack Tube Pressure sensor
                - Fixed line endings in sensors.cpp
                - Fixed bug in VelocityToVolumetric method
```

This is a phenomenal amount of work to get done in such a short time.

Once again, a BIG thanks to all those who helped.

**Comment by DeeEmm** [comment]
OK RC9 now released.

This discussion now locked.

Please upgrade to current version and direct comments to the new discussion thread for RC9 - https://github.com/DeeEmm/DIY-Flow-Bench/discussions/367
